File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/digitaltwin/element/ElementCollectionType.java
package org.opcfoundation.webserver.digitaltwin.element;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.Argument;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.*;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjectTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.callback.ElementCollectionCallback;
import org.opcfoundation.webserver.service.message.digitaltwin.*;
import org.opcfoundation.webserver.service.message.reactiveobject.*;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.common.UaChildId;
import org.opcfoundation.webserver.types.common.UaObjectId;
import org.opcfoundation.webserver.types.common.UaReferenceDescriptor;
import org.opcfoundation.webserver.types.digitaltwin.ObjectServiceContext;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public abstract class ElementCollectionType extends ElementType implements ElementCollectionCallback {
    public ElementCollectionType(String typeId,
                               LocalizedText displayName,
                               DigitalTwinSpace twinSpace)
    {
        super(
                typeId,
                displayName,
                UaObjectTypes.ElementCollectionType,
                twinSpace);
    }

    public UaVariable addPropertyElement(
            String        name,
            LocalizedText displayName,
            LocalizedText description,
            UaDataType    dataType,
            boolean       writable)
    {
        return addPropertyElement(
                name,
                displayName,
                description,
                dataType,
                writable,
                false,
                -1,
                UaVariableTypes.PropertyType,
                true);
    }

    public UaVariable addPropertyElement(
            String                         name,
            LocalizedText                  displayName,
            LocalizedText                  description,
            UaDataType                     dataType,
            boolean                        writable,
            boolean                        historizing,
            @Nullable Integer              valueRank,
            @Nullable UaVariableType       variableType,
            boolean                        mandatory)
    {
        UaVariable newVariable = addVariableNode(name, displayName, dataType,writable, historizing, valueRank, variableType);
        if (description.isNotNull()) newVariable.setDescription(description);
        newVariable.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newVariable;
    }

    public void addSubElementOfProperty(UaVariable property, String subElementName, Variant value)
    {
        UaVariable subElement = property.addMember(subElementName);
        if (null != subElement)
        {
            subElement.setValue(value);
            nodeManager.addNode(subElement);
        }
    }

    public UaMethod addOperationElement(
            String                   name,
            LocalizedText            displayName,
            LocalizedText            description,
            @Nullable List<Argument> inputArguments,
            @Nullable List<Argument> outputArguments,
            boolean                  mandatory)
    {
        UaMethod newMethod = addMethodNode(name, displayName, inputArguments, outputArguments);
        if (description.isNotNull()) newMethod.setDescription(description);
        newMethod.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newMethod;
    }

    public UaObject addElementCollection(
            ElementCollectionType type,
            String                name,
            LocalizedText         displayName,
            LocalizedText         description,
            boolean               mandatory)
    {
        UaObject newObject = addObjectNode(name, displayName, type);
        if (description.isNotNull()) newObject.setDescription(description);
        newObject.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newObject;
    }

    public UaObject addReferenceElement(
            ReferenceElementType type,
            String               name,
            LocalizedText        displayName,
            LocalizedText        description,
            boolean              mandatory)
    {
        UaObject newObject = addObjectNode(name, displayName, type);
        if (description.isNotNull()) newObject.setDescription(description);
        newObject.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newObject;
    }

    public UaObject addEventElement(
            EventElementType type,
            String           name,
            LocalizedText    displayName,
            LocalizedText    description,
            boolean          mandatory)
    {
        UaObject newObject = addObjectNode(name, displayName, type);
        if (description.isNotNull()) newObject.setDescription(description);
        newObject.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newObject;
    }

    public UaObject addElementList(
            ElementListType type,
            String          name,
            LocalizedText   displayName,
            LocalizedText   description,
            boolean         mandatory)
    {
        UaObject newObject = addObjectNode(name, displayName, type);
        if (description.isNotNull()) newObject.setDescription(description);
        newObject.setModellingRule((mandatory) ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newObject;
    }

    @Override
    public final CompletableFuture<ReadObjectAttributeResponse> onReadObjectAttributes(ReadObjectAttributeRequest request)
    {
        UaObject instanceDeclaration = request.getObjectId().getInstance();

        if (null == instanceDeclaration)
        {
            ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
            GetDescriptorRequest getDescriptorRequest = new GetDescriptorRequest(context);

            return onGetDescriptor(getDescriptorRequest).thenApply(response -> {
                return new ReadObjectAttributeResponse(request.getObjectId().getId(), response.getDisplayName(), response.getDescription());
            });
        } else {
            return CompletableFuture.completedFuture(
                    new ReadObjectAttributeResponse(
                            instanceDeclaration.browseName(),
                            instanceDeclaration.displayName(),
                            instanceDeclaration.description()));
        }
    }

    @Override
    public final CompletableFuture<BrowseObjectResponse> onBrowseObjectChildren(BrowseObjectRequest request)
    {
        NodeId referenceTypeId = request.getBrowseDescription().getReferenceTypeId();

        List<UaInstanceNode> members = getMembers();
        final List<UaInstanceNode> membersToReturn = new ArrayList<>();

        for (UaInstanceNode item: members)
        {
            if (item.nodeClass() == NodeClass.Object)
            {
                if (!request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK)) continue;
            }

            if (item.nodeClass() == NodeClass.Method)
            {
                if (!request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK)) continue;
            }

            if (item.nodeClass() == NodeClass.Variable)
            {
                if (!request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK)) continue;

                NodeId variableTypeId = ((UaVariable) item).typeDefinition().nodeId();
                if (referenceTypeId.equals(NodeIds.HasProperty))
                {
                    if (!variableTypeId.equals(NodeIds.PropertyType)) continue;
                } else if (referenceTypeId.equals(NodeIds.HasComponent)) {
                    if (variableTypeId.equals(NodeIds.PropertyType)) continue;
                }
            }

            membersToReturn.add(item);
        }

        if (membersToReturn.isEmpty()) return CompletableFuture.completedFuture(new BrowseObjectResponse(new ArrayList<>(), false));

        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        GetElementsRequest getElementsRequest = new GetElementsRequest(context);

        return onGetElements(getElementsRequest).
                thenApply(response -> processBrowseObjectChildrenResponse(request.getObjectId(),membersToReturn, response));
    }

    @Override
    public final CompletableFuture<BrowseMemberResponse> onBrowseMemberChildren(BrowseMemberRequest request)
    {
        NodeId referenceTypeId = request.getBrowseDescription().getReferenceTypeId();

        UaInstanceNode member = getMember(request.getChildId());
        if (null == member || NodeClass.Object == member.nodeClass()) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();
        List<UaInstanceNode> members = member.getMembers();

        for (UaInstanceNode item: members)
        {
            if (item.nodeClass() != NodeClass.Variable) continue;

            NodeId variableTypeId = ((UaVariable)item).typeDefinition().nodeId();

            if (referenceTypeId.equals(NodeIds.HasProperty))
            {
                if (!variableTypeId.equals(NodeIds.PropertyType)) continue;
            } else if (referenceTypeId.equals(NodeIds.HasComponent)) {
                if (variableTypeId.equals(NodeIds.PropertyType)) continue;
            }

            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    item.browseName(),
                    item,
                    variableTypeId.equals(NodeIds.PropertyType) ? NodeIds.HasProperty : NodeIds.HasComponent,
                    true);

            childDescriptors.add(descriptor);
        }

        return CompletableFuture.completedFuture(new BrowseMemberResponse(childDescriptors));
    }

    @Override
    public final CompletableFuture<ReadMemberAttributeResponse> onReadMemberAttributes(ReadMemberAttributeRequest request)
    {
        return CompletableFuture.supplyAsync(() -> {
            UaInstanceNode memberNode = getMember(request.getChildId().getId());

            if (null != memberNode && null != request.getChildId().getSubElementName())
            {
                memberNode = memberNode.getMember(request.getChildId().getSubElementName());
            }

            if (null == memberNode) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            return new ReadMemberAttributeResponse(memberNode);
        });
    }

    @Override
    public final CompletableFuture<ReadVariableValueResponse> onReadVariablesValue(ReadVariableValueRequest request)
    {
        final Set<String> propertyNames = new HashSet<>();
        final Set<UaChildId> subElementNames = new HashSet<>();

        for (UaChildId item: request.getVariableIds())
        {
            if (null == item.getSubElementName())
            {
                propertyNames.add(item.getId());
            } else {
                subElementNames.add(item);
            }
        }

        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        ReadPropertyValuesRequest readPropertyValuesRequest = new ReadPropertyValuesRequest(
                context,
                propertyNames);

        if (propertyNames.isEmpty())
        {
            return CompletableFuture.completedFuture(processReadVariableValueResponse(subElementNames, new ReadPropertyValuesResponse()));
        }

        return onReadPropertyValues(readPropertyValuesRequest).
                thenApply(readPropertyValuesResponse -> {
                    return processReadVariableValueResponse(subElementNames, readPropertyValuesResponse); });
    }

    @Override
    public final CompletableFuture<WriteVariableValueResponse> onWriteVariablesValue(WriteVariableValueRequest request)
    {
        Map<String, Variant> elementValues = new HashMap<>();

        for (Map.Entry<UaChildId, Variant> item : request.getVariableValues().entrySet())
        {
            UaChildId childId = item.getKey();
            if (null != childId.getSubElementName()) continue;

            UaNode memberNode = getMember(childId.getId());
            if (null == memberNode || NodeClass.Variable != memberNode.nodeClass()) continue;

            UaVariable variableToWrite = (UaVariable) memberNode;
            if ((AccessLevel.CurrentWrite.getValue() & variableToWrite.accessLevel()) == 0) continue;

            elementValues.put(childId.getId(), item.getValue());
        }

        if (elementValues.isEmpty())
        {
            return CompletableFuture.completedFuture(new WriteVariableValueResponse());
        }

        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        WritePropertyValuesRequest writePropertyValuesRequest = new WritePropertyValuesRequest(
                context,
                elementValues);

        return onWritePropertyValues(writePropertyValuesRequest)
                .thenApply(response -> {
                    return new WriteVariableValueResponse(response.getResults());
                });
    }

    @Override
    public CompletableFuture<MethodCallResponse> onMethodCall(MethodCallRequest request)
    {
        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        InvokeOperationRequest invokeOperationRequest = new InvokeOperationRequest(
                context,
                request.getMethodName(),
                request.getInputArguments());

        return onInvokeOperation(invokeOperationRequest)
                .thenApply(response -> {
                    return new MethodCallResponse(response.getOutputArguments());
                });
    }

    private BrowseObjectResponse processBrowseObjectChildrenResponse(
            UaObjectId objectId,
            List<UaInstanceNode> members,
            GetElementsResponse response)
    {
        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();

        for (UaInstanceNode item: members)
        {
            if (item.modellingRule() == UaModellingRule.Optional &&
                    !response.getElementNames().contains(item.browseName())) continue;

            NodeId referenceType = (item.nodeClass() == NodeClass.Variable &&
                    ((UaVariable) item).typeDefinition().nodeId().equals(NodeIds.PropertyType)) ? NodeIds.HasProperty : NodeIds.HasComponent;

            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    (item.nodeClass() == NodeClass.Object) ? objectId.getId() : item.browseName(),
                    item,
                    referenceType,
                    true);

            childDescriptors.add(descriptor);
        }

        return new BrowseObjectResponse(childDescriptors, false);
    }

    private ReadVariableValueResponse processReadVariableValueResponse(
            Set<UaChildId> subElementNames,
            ReadPropertyValuesResponse response)
    {
        ReadVariableValueResponse readVariableValueResponse = new ReadVariableValueResponse(response.getResults());
        Map<UaChildId, DataValue> results = response.getResults();

        for (UaChildId item: subElementNames)
        {
            if (null == item.getSubElementName()) continue;

            UaInstanceNode node = getMember(item.getId());
            if (null == node || NodeClass.Variable != node.nodeClass()) continue;

            UaInstanceNode subElementNode = node.getMember(item.getSubElementName());
            if (null == subElementNode || NodeClass.Variable != subElementNode.nodeClass()) continue;

            results.put(
                    new UaChildId(item.getId(), item.getSubElementName()),
                    new DataValue(((UaVariable)subElementNode).value(), StatusCode.GOOD, null, null));
        }

        return readVariableValueResponse;
    }

    @Override
    public final List<UaInstanceNode> getElements()
    {
        return getMembers();
    }
}

File: opcua-webservice-node/source/src/server/digitaltwin/element/ElementCollectionType.ts
import { NodeClass, StatusCodes } from "opcua-webapi";
import { ReferenceTypeIds, UaAccessLevel, UaArgument, UaDataValue, UaError, UaLocalizedText, UaNodeId, UaVariant, makeUaStatusCode } from "opcua-webapi-ts";
import { UaDataType } from "../../addressspace/nodes/UaDataType";
import { UaInstanceNode } from "../../addressspace/nodes/UaInstanceNode";
import { UaMethod } from "../../addressspace/nodes/UaMethod";
import { UaModellingRule } from "../../addressspace/nodes/UaModellingRule";
import { UaObject } from "../../addressspace/nodes/UaObject";
import { UaVariable } from "../../addressspace/nodes/UaVariable";
import type { UaVariableType } from "../../addressspace/nodes/UaVariableType";
import { UaObjectTypes, UaVariableTypes } from "../../addressspace/nodes/builtin";
import { ElementCollectionCallback } from "../callback/ElementCollectionCallback";

import {
    BrowseMemberRequest,
    BrowseMemberResponse,
    BrowseObjectRequest,
    BrowseObjectResponse,
    GetDescriptorRequest,
    GetDescriptorResponse,
    GetElementsRequest,
    GetElementsResponse,
    InvokeOperationRequest,
    InvokeOperationResponse,
    MethodCallRequest,
    MethodCallResponse,
    ReadHistoryDataRequest,
    ReadHistoryDataResponse,
    ReadMemberAttributeRequest,
    ReadMemberAttributeResponse,
    ReadObjectAttributeRequest,
    ReadObjectAttributeResponse,
    ReadPropertyHistoryValuesRequest,
    ReadPropertyHistoryValuesResponse,
    ReadPropertyValuesRequest,
    ReadPropertyValuesResponse,
    ReadVariableValueRequest,
    ReadVariableValueResponse,
    WritePropertyValuesRequest,
    WritePropertyValuesResponse,
    WriteVariableValueRequest,
    WriteVariableValueResponse
} from "../../service/message";
import { UaBrowseAdditionalInfo, UaChildId, UaObjectId, UaReferenceDescriptor } from "../../types";
import { ObjectServiceContext } from "../../types/digitaltwin/ObjectServiceContext";
import { ElementListType } from "./ElementListType";
import { ElementType } from "./ElementType";
import { EventElementType } from "./EventElementType";
import { ReferenceElementType } from "./ReferenceElementType";
import { DigitalTwinSpace } from "../DigitalTwinSpace";

export abstract class ElementCollectionType extends ElementType implements ElementCollectionCallback {
    constructor(
        typeId: string,
        displayName: UaLocalizedText,
        twinSpace: DigitalTwinSpace) {
        super(typeId, displayName, UaObjectTypes.ElementCollectionType, twinSpace);
    }

    // Override to read property values
    abstract onReadPropertyValues(request: ReadPropertyValuesRequest): Promise<ReadPropertyValuesResponse>;
    
    // Override to update property value
    async onWritePropertyValues(request: WritePropertyValuesRequest): Promise<WritePropertyValuesResponse> {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }
    
    // Override to process a method call
    async onInvokeOperation(request: InvokeOperationRequest): Promise<InvokeOperationResponse> {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    // Override to read history data
    async onReadPropertyHistoryValues(request: ReadPropertyHistoryValuesRequest): Promise<ReadPropertyHistoryValuesResponse>
    {        
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    // Can be override to provide customized descriptor
    async onGetDescriptor(request: GetDescriptorRequest): Promise<GetDescriptorResponse>
    {
        const instance = request.context.objectId.instance;

        if (instance === null) {
            return new GetDescriptorResponse("NotImplemented");
        }

        return new GetDescriptorResponse(instance.displayName, instance.description);
    }

    // Can be override to provide customized elements
    async onGetElements(request: GetElementsRequest): Promise<GetElementsResponse>
    {        
        let response = new GetElementsResponse();
        for (const item of this.getMembers()) {
            response.elementNames.add(item.browseName);
        }
        return response;
    }

    addPropertyElement(
        name: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        dataType: UaDataType,
        writable: boolean,
    ): UaVariable;
    addPropertyElement(
        name: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        dataType: UaDataType,
        writable: boolean,
        historizing = false,
        valueRank: number = -1,
        variableType: UaVariableType = UaVariableTypes.PropertyType,
        mandatory = true): UaVariable {
        const newVariable = this.addVariableNode(name, displayName, dataType, writable, historizing, valueRank, variableType);
        if (description.text.length > 0) {
            newVariable.description = description;
        }
        newVariable.setModellingRule(mandatory ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newVariable;
    }

    addSubElementOfProperty(property: UaVariable, subElementName: string, value: UaVariant): void {
        const subElement = property.addMemberByName(subElementName);
        if (subElement !== null) {
            subElement.value = value;
            this.nodeManager.addNode(subElement);
        }
    }

    addOperationElement(
        name: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        inputArguments: Array<UaArgument> | null,
        outputArguments: Array<UaArgument> | null,
        mandatory: boolean,
    ): UaMethod {
        const newMethod = this.addMethodNode(name, displayName, inputArguments, outputArguments);
        if (description.text.length > 0) {
            newMethod.description = description;
        }
        newMethod.setModellingRule(mandatory ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newMethod;
    }

    addElementCollection(type: ElementCollectionType, name: string, displayName: UaLocalizedText, description: UaLocalizedText, mandatory: boolean): UaObject {
        return this.addChildObject(type, name, displayName, description, mandatory);
    }

    addReferenceElement(type: ReferenceElementType, name: string, displayName: UaLocalizedText, description: UaLocalizedText, mandatory: boolean): UaObject {
        return this.addChildObject(type, name, displayName, description, mandatory);
    }

    addEventElement(type: EventElementType, name: string, displayName: UaLocalizedText, description: UaLocalizedText, mandatory: boolean): UaObject {
        return this.addChildObject(type, name, displayName, description, mandatory);
    }

    addElementList(type: ElementListType, name: string, displayName: UaLocalizedText, description: UaLocalizedText, mandatory: boolean): UaObject {
        return this.addChildObject(type, name, displayName, description, mandatory);
    }

    // Implementation of parent type methods, don't override
    override async onReadObjectAttributes(request: ReadObjectAttributeRequest): Promise<ReadObjectAttributeResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetDescriptor(new GetDescriptorRequest(context));
        return new ReadObjectAttributeResponse(request.objectId.id, response.displayName, response.description);
    }

    // Implementation of parent type methods, don't override 
    override async onBrowseObjectChildren(request: BrowseObjectRequest): Promise<BrowseObjectResponse> {
        const referenceTypeId = request.browseDescription.referenceTypeId;
        const membersToReturn: Array<UaInstanceNode> = [];

        for (const item of this.getMembers()) {
            if (item.nodeClass === NodeClass.Object && !request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK)) {
                continue;
            }

            if (item.nodeClass === NodeClass.Method && !request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK)) {
                continue;
            }

            if (item.nodeClass === NodeClass.Variable) {
                if (!request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK)) {
                    continue;
                }

                const variableTypeId = (item as UaVariable).typeDefinition.nodeId;
                if (variableTypeId.equal(UaVariableTypes.PropertyType.nodeId)) {
                    if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasComponent))) {
                        continue;
                    }
                } else if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasProperty))) {
                    continue;
                }
            }

            membersToReturn.push(item);
        }

        if (membersToReturn.length === 0) {
            return new BrowseObjectResponse([], false);
        }

        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetElements(new GetElementsRequest(context));
        return this.processBrowseObjectChildrenResponse(request.objectId, membersToReturn, response);
    }

    // Implementation of parent type methods, don't override 
    override async onBrowseMemberChildren(request: BrowseMemberRequest): Promise<BrowseMemberResponse> {
        const referenceTypeId = request.browseDescription.referenceTypeId;
        const member = this.getMember(request.childId);
        if (member === null || member.nodeClass === NodeClass.Object) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
        }

        const childDescriptors: Array<UaReferenceDescriptor> = [];
        for (const item of member.getMembers()) {
            if (item.nodeClass !== NodeClass.Variable) {
                continue;
            }

            const variableTypeId = (item as UaVariable).typeDefinition.nodeId;
            if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasProperty))) {
                if (!variableTypeId.equal(UaVariableTypes.PropertyType.nodeId)) {
                    continue;
                }
            } else if (referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HasComponent))) {
                if (variableTypeId.equal(UaVariableTypes.PropertyType.nodeId)) {
                    continue;
                }
            }

            childDescriptors.push(
                UaReferenceDescriptor.fromInstanceDeclaration(
                    item.browseName,
                    item,
                    variableTypeId.equal(UaVariableTypes.PropertyType.nodeId)
                        ? UaNodeId.from(ReferenceTypeIds.HasProperty)
                        : UaNodeId.from(ReferenceTypeIds.HasComponent),
                    true,
                ),
            );
        }

        return new BrowseMemberResponse(childDescriptors);
    }

    // Implementation of parent type methods, don't override 
    override async onReadMemberAttributes(request: ReadMemberAttributeRequest): Promise<ReadMemberAttributeResponse> {
        let memberNode = this.getMember(request.childId.id);

        if (memberNode !== null && request.childId.subElementName !== null) {
            memberNode = memberNode.getMember(request.childId.subElementName);
        }

        if (memberNode === null) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
        }

        return ReadMemberAttributeResponse.fromInstanceDeclaration(memberNode);
    }

    // Implementation of parent type methods, don't override 
    override async onReadVariablesValue(request: ReadVariableValueRequest): Promise<ReadVariableValueResponse> {
        const propertyNames = new Set<string>();
        const subElementNames = new Set<UaChildId>();

        for (const item of request.variableIds) {
            const childId = this.toChildId(item);
            if (childId.subElementName === null) {
                propertyNames.add(childId.id);
            } else {
                subElementNames.add(childId);
            }
        }

        const context = new ObjectServiceContext(request.objectId);
        const readResponse = propertyNames.size === 0
            ? new ReadPropertyValuesResponse()
            : await this.onReadPropertyValues(new ReadPropertyValuesRequest(context, propertyNames));

        return this.processReadVariableValueResponse(subElementNames, readResponse);
    }

    // Implementation of parent type methods, don't override 
    override async onWriteVariablesValue(request: WriteVariableValueRequest): Promise<WriteVariableValueResponse> {
        const elementValues = new Map<string, UaVariant>();

        for (const [childIdText, value] of request.variableValues) {
            const childId = this.toChildId(childIdText);
            if (childId.subElementName !== null) {
                continue;
            }

            const memberNode = this.getMember(childId.id);
            if (memberNode === null || memberNode.nodeClass !== NodeClass.Variable) {
                continue;
            }

            const variableToWrite = memberNode as UaVariable;
            if ((UaAccessLevel.CurrentWrite & variableToWrite.accessLevel) === 0) {
                continue;
            }

            elementValues.set(childId.id, value);
        }

        if (elementValues.size === 0) {
            return new WriteVariableValueResponse();
        }

        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onWritePropertyValues(new WritePropertyValuesRequest(context, elementValues));
        return new WriteVariableValueResponse(response.results);
    }

    // Implementation of parent type methods, don't override 
    override async onMethodCall(request: MethodCallRequest): Promise<MethodCallResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onInvokeOperation(
            new InvokeOperationRequest(context, request.methodName, request.inputArguments),
        );

        return new MethodCallResponse(response.outputArguments);
    }

    // Implementation of parent type methods, don't override 
    override async onReadHistoryData(request: ReadHistoryDataRequest): Promise<ReadHistoryDataResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onReadPropertyHistoryValues(
            new ReadPropertyHistoryValuesRequest(context, request.childId.id, request.details),
        );

        return this.processReadHistoryValueResponse(response);
    }
    
    private addChildObject(
        type: ElementType,
        name: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        mandatory: boolean,
    ): UaObject {
        const newObject = this.addObjectNode(name, displayName, type);
        if (description.text.length > 0) {
            newObject.description = description;
        }
        newObject.setModellingRule(mandatory ? UaModellingRule.Mandatory : UaModellingRule.Optional);
        return newObject;
    }

    private processBrowseObjectChildrenResponse(
        objectId: UaObjectId,
        members: Array<UaInstanceNode>,
        response: GetElementsResponse): BrowseObjectResponse {
        const childDescriptors: Array<UaReferenceDescriptor> = [];

        for (const item of members) {
            if (item.modellingRule === UaModellingRule.Optional && !response.elementNames.has(item.browseName)) {
                continue;
            }

            const referenceType = item.nodeClass === NodeClass.Variable &&
                (item as UaVariable).typeDefinition.nodeId.equal(UaVariableTypes.PropertyType.nodeId)
                ? UaNodeId.from(ReferenceTypeIds.HasProperty)
                : UaNodeId.from(ReferenceTypeIds.HasComponent);

            childDescriptors.push(
                UaReferenceDescriptor.fromInstanceDeclaration(
                    item.nodeClass === NodeClass.Object ? objectId.id : item.browseName,
                    item,
                    referenceType,
                    true,
                ),
            );
        }

        return new BrowseObjectResponse(childDescriptors, false);
    }

    private processReadVariableValueResponse(
        subElementNames: Set<UaChildId>,
        response: ReadPropertyValuesResponse,
    ): ReadVariableValueResponse {
        const readVariableValueResponse = new ReadVariableValueResponse();

        for (const [key, value] of response.results) {
            readVariableValueResponse.results.set(key, value);
        }

        for (const item of subElementNames) {
            if (item.subElementName === null) {
                continue;
            }

            const node = this.getMember(item.id);
            if (node === null || node.nodeClass !== NodeClass.Variable) {
                continue;
            }

            const subElementNode = node.getMember(item.subElementName);
            if (subElementNode === null || subElementNode.nodeClass !== NodeClass.Variable) {
                continue;
            }

            let subVariable = subElementNode as UaVariable;
            readVariableValueResponse.results.set(
                new UaChildId(item.id, item.subElementName).toString(),
                new UaDataValue(subVariable.value, makeUaStatusCode(StatusCodes.Good)),
            );
        }

        return readVariableValueResponse;
    }

    private processReadHistoryValueResponse(response: ReadPropertyHistoryValuesResponse): ReadHistoryDataResponse {
        return new ReadHistoryDataResponse(response.dataValues, response.containsMoreData);
    }

    private toChildId(value: string): UaChildId {
        const [id, subElementName] = value.split("#", 2);
        return new UaChildId(id, subElementName ?? null);
    }
}